edb/pgsql/compiler/relctx.py

```diff
@@ -1640,49 +1641,104 @@ def range_for_material_objtype(
             rvar = rvar_for_rel(
                 sctx.rel, lateral=lateral, typeref=typeref, ctx=sctx)
 
-    # When we are compiling a query for EXPLAIN, expand out type references
-    # to an explicit union of all the types, rather than relying on the
-    # inheritance views. This allows postgres to actually give us back the
-    # alias names that we use for relations, which we use to track which
-    # parts of the query are being referred to.
     elif (
-        ctx.env.expand_inhviews
-        and include_descendants
+        include_descendants
         and not for_mutation
         and typeref.children is not None
 
         # HACK: This is a workaround for #4491
         and typeref.name_hint.module not in {'cfg', 'sys'}
     ):
-        ops = []
-        typerefs = [typeref, *irtyputils.get_typeref_descendants(typeref)]
-        all_abstract = all(subref.is_abstract for subref in typerefs)
-        for subref in typerefs:
-            if subref.is_abstract and not all_abstract:
-                continue
-            rvar = range_for_material_objtype(
-                subref, path_id, lateral=lateral,
-                include_descendants=False,
-                include_overlays=False,
-                ignore_rewrites=ignore_rewrites,  # XXX: Is this right?
+        if (
+            # When we are compiling a query for EXPLAIN, expand out type
+            # references to an explicit union of all the types, rather than
+            # using a CTE. This allows postgres to actually give us back the
+            # alias names that we use for relations, which we use to track which
+            # parts of the query are being referred to.
+            not ctx.env.use_inheritance_ctes
+
+            # Don't use CTEs if there is no inheritance. (ie. There is only a
+            # single material type)
+            or len(typeref.children) == 0
```

Comment:
I think the `== 0` case should maybe be folded into the above condition, so we don't add an extra select around it for no reason